File: ios/Sources/RynBridgeCrypto/DefaultCryptoProvider.swift
import Foundation
import CryptoKit
import RynBridge

public final class DefaultCryptoProvider: CryptoProvider, @unchecked Sendable {
    private var privateKey: Curve25519.KeyAgreement.PrivateKey
    private var symmetricKey: SymmetricKey?
    private var keyCreatedAt: Date

    public init() {
        self.privateKey = Curve25519.KeyAgreement.PrivateKey()
        self.keyCreatedAt = Date()
    }

    public func generateKeyPair() async throws -> String {
        privateKey = Curve25519.KeyAgreement.PrivateKey()
        keyCreatedAt = Date()
        symmetricKey = nil
        return privateKey.publicKey.rawRepresentation.base64EncodedString()
    }

    public func performKeyExchange(remotePublicKey: String) async throws -> Bool {
        guard let remoteKeyData = Data(base64Encoded: remotePublicKey) else {
            throw RynBridgeError(code: .invalidMessage, message: "Invalid base64 public key")
        }
        let remoteKey = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: remoteKeyData)
        let sharedSecret = try privateKey.sharedSecretFromKeyAgreement(with: remoteKey)
        symmetricKey = sharedSecret.hkdfDerivedSymmetricKey(
            using: SHA256.self,
            salt: Data("RynBridge".utf8),
            sharedInfo: Data(),
            outputByteCount: 32
        )
        return true
    }

    public func encrypt(data: String, associatedData: String?) async throws -> EncryptResult {
        guard let key = symmetricKey else {
            throw RynBridgeError(code: .unknown, message: "Key exchange not performed. Call performKeyExchange first.")
        }
        let plaintext = Data(data.utf8)
        let aad = associatedData.map { Data($0.utf8) }
        let sealedBox: AES.GCM.SealedBox
        if let aad {
            sealedBox = try AES.GCM.seal(plaintext, using: key, authenticating: aad)
        } else {
            sealedBox = try AES.GCM.seal(plaintext, using: key)
        }
        return EncryptResult(
            ciphertext: sealedBox.ciphertext.base64EncodedString(),
            iv: sealedBox.nonce.withUnsafeBytes { Data($0) }.base64EncodedString(),
            tag: sealedBox.tag.base64EncodedString()
        )
    }

    public func decrypt(ciphertext: String, iv: String, tag: String, associatedData: String?) async throws -> String {
        guard let key = symmetricKey else {
            throw RynBridgeError(code: .unknown, message: "Key exchange not performed. Call performKeyExchange first.")
        }
        guard let ciphertextData = Data(base64Encoded: ciphertext),
              let ivData = Data(base64Encoded: iv),
              let tagData = Data(base64Encoded: tag) else {
            throw RynBridgeError(code: .invalidMessage, message: "Invalid base64 encoded data")
        }
        let nonce = try AES.GCM.Nonce(data: ivData)
        let sealedBox = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertextData, tag: tagData)
        let decryptedData: Data
        let aad = associatedData.map { Data($0.utf8) }
        if let aad {
            decryptedData = try AES.GCM.open(sealedBox, using: key, authenticating: aad)
        } else {
            decryptedData = try AES.GCM.open(sealedBox, using: key)
        }
        guard let plaintext = String(data: decryptedData, encoding: .utf8) else {
            throw RynBridgeError(code: .unknown, message: "Failed to decode decrypted data as UTF-8")
        }
        return plaintext
    }

    public func getStatus() async throws -> CryptoStatus {
        let formatter = ISO8601DateFormatter()
        return CryptoStatus(
            initialized: symmetricKey != nil,
            keyCreatedAt: formatter.string(from: keyCreatedAt),
            algorithm: "AES-256-GCM"
        )
    }

    public func rotateKeys() async throws -> String {
        privateKey = Curve25519.KeyAgreement.PrivateKey()
        keyCreatedAt = Date()
        symmetricKey = nil
        return privateKey.publicKey.rawRepresentation.base64EncodedString()
    }
}

File: android/crypto/src/main/kotlin/com/devryner/rynbridge/crypto/DefaultCryptoProvider.kt
package com.devryner.rynbridge.crypto

import com.devryner.rynbridge.core.ErrorCode
import com.devryner.rynbridge.core.RynBridgeError
import java.security.KeyFactory
import java.security.KeyPairGenerator
import java.security.PublicKey
import java.security.interfaces.ECPublicKey
import java.security.spec.ECGenParameterSpec
import java.security.spec.X509EncodedKeySpec
import java.time.Instant
import java.util.Base64
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.SecretKey
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

class DefaultCryptoProvider : CryptoProvider {

    private var keyPair = generateECKeyPair()
    private var symmetricKey: SecretKey? = null
    private var keyCreatedAt: Instant = Instant.now()

    private fun generateECKeyPair(): java.security.KeyPair {
        val generator = KeyPairGenerator.getInstance("EC")
        generator.initialize(ECGenParameterSpec("secp256r1"))
        return generator.generateKeyPair()
    }

    override suspend fun generateKeyPair(): String {
        keyPair = generateECKeyPair()
        keyCreatedAt = Instant.now()
        symmetricKey = null
        return Base64.getEncoder().encodeToString(keyPair.public.encoded)
    }

    override suspend fun performKeyExchange(remotePublicKey: String): Boolean {
        val remoteKeyBytes = Base64.getDecoder().decode(remotePublicKey)
        val keyFactory = KeyFactory.getInstance("EC")
        val remoteKey = keyFactory.generatePublic(X509EncodedKeySpec(remoteKeyBytes))

        val agreement = KeyAgreement.getInstance("ECDH")
        agreement.init(keyPair.private)
        agreement.doPhase(remoteKey, true)
        val sharedSecret = agreement.generateSecret()

        // Derive 256-bit key from shared secret using first 32 bytes of SHA-256
        val digest = java.security.MessageDigest.getInstance("SHA-256")
        digest.update("RynBridge".toByteArray())
        digest.update(sharedSecret)
        val derivedKey = digest.digest()
        symmetricKey = SecretKeySpec(derivedKey, "AES")
        return true
    }

    override suspend fun encrypt(data: String, associatedData: String?): EncryptResult {
        val key = symmetricKey
            ?: throw RynBridgeError(code = ErrorCode.UNKNOWN, message = "Key exchange not performed. Call performKeyExchange first.")

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key)

        associatedData?.let { cipher.updateAAD(it.toByteArray()) }

        val ciphertext = cipher.doFinal(data.toByteArray())
        val iv = cipher.iv
        val tagLength = 16
        val encryptedData = ciphertext.copyOfRange(0, ciphertext.size - tagLength)
        val tag = ciphertext.copyOfRange(ciphertext.size - tagLength, ciphertext.size)

        return EncryptResult(
            ciphertext = Base64.getEncoder().encodeToString(encryptedData),
            iv = Base64.getEncoder().encodeToString(iv),
            tag = Base64.getEncoder().encodeToString(tag)
        )
    }

    override suspend fun decrypt(ciphertext: String, iv: String, tag: String, associatedData: String?): String {
        val key = symmetricKey
            ?: throw RynBridgeError(code = ErrorCode.UNKNOWN, message = "Key exchange not performed. Call performKeyExchange first.")

        val ciphertextBytes = Base64.getDecoder().decode(ciphertext)
        val ivBytes = Base64.getDecoder().decode(iv)
        val tagBytes = Base64.getDecoder().decode(tag)

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        val spec = GCMParameterSpec(128, ivBytes)
        cipher.init(Cipher.DECRYPT_MODE, key, spec)

        associatedData?.let { cipher.updateAAD(it.toByteArray()) }

        val combined = ciphertextBytes + tagBytes
        val plaintext = cipher.doFinal(combined)
        return String(plaintext)
    }

    override suspend fun getStatus(): CryptoStatus {
        return CryptoStatus(
            initialized = symmetricKey != null,
            keyCreatedAt = keyCreatedAt.toString(),
            algorithm = "AES-256-GCM"
        )
    }

    override suspend fun rotateKeys(): String {
        keyPair = generateECKeyPair()
        keyCreatedAt = Instant.now()
        symmetricKey = null
        return Base64.getEncoder().encodeToString(keyPair.public.encoded)
    }
}
